File: problems/2131_longest_palindrome_by_concatenating_two_letter_words/solution.py
class Solution:
    def longestPalindrome(self, words: list[str]) -> int:
        from collections import Counter
        freq = Counter(words)
        length = 0
        has_center = False
        for word, count in freq.items():
            rev = word[::-1]
            if word == rev:
                length += (count // 2) * 4
                if count % 2 == 1:
                    has_center = True
            elif word < rev and rev in freq:
                length += min(count, freq[rev]) * 4
        if has_center:
            length += 2
        return length

File: problems/2131_longest_palindrome_by_concatenating_two_letter_words/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> freq;
        for (auto& w : words) freq[w]++;
        int length = 0;
        bool hasCenter = false;
        for (auto& [word, count] : freq) {
            string rev(word.rbegin(), word.rend());
            if (word == rev) {
                length += (count / 2) * 4;
                if (count % 2 == 1) hasCenter = true;
            } else if (word < rev && freq.count(rev)) {
                length += min(count, freq[rev]) * 4;
            }
        }
        if (hasCenter) length += 2;
        return length;
    }
};
